tarantool_test.go

```diff
@@ -485,6 +687,102 @@ func BenchmarkSQLSerial(b *testing.B) {
 
 ///////////////////
 
+func TestClientDoWithContext(t *testing.T) {
+	var err error
+	var conn *Connection
+
+	conn, err = Connect(server, opts)
+	if err != nil {
+		t.Errorf("Failed to connect: %s", err.Error())
+		return
+	}
+	if conn == nil {
+		t.Errorf("conn is nil after Connect")
+		return
+	}
+	defer conn.Close()
+
+	// The code prepares data.
+	for i := 1010; i < 1020; i++ {
+		_, err = conn.Replace(spaceName, []interface{}{uint(i), fmt.Sprintf("val %d", i), "bla"})
+		if err != nil {
+			t.Errorf("Failed to Replace: %s", err.Error())
+		}
+	}
+
+	req := NewSelectRequest(spaceNo).
+		Index(indexNo).
+		Limit(1).
+		Iterator(IterEq).
+		Key([]interface{}{uint(1010)})
+
+	ctx, cancel := context.WithCancel(context.Background())
+	cancel()
+	_, err = conn.Do(ctx, req)
+	if err.Error() != "context is done" {
+		t.Errorf("wrong failure message")
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), time.Nanosecond)
+	cancel()
+	_, err = conn.Do(ctx, req)
+	if err.Error() != "context is done" {
+		t.Errorf("wrong failure message")
+	}
+
+	_, err = conn.Do(context.Background(), req)
+	if err != nil {
+		t.Errorf("failed DoWithContext with Background context, %s", err)
+	}
+
+	_, err = conn.Do(context.TODO(), req)
+	if err != nil {
+		t.Errorf("failed DoWithContext with TODO context, %s", err)
+	}
+
+	_, err = conn.Do(nil, req) //nolint
+	if err.Error() != "passed nil context" {
+		t.Errorf("failed DoWithContext with nil context, %s", err)
+	}
+}
+
+func TestFutureWithContext(t *testing.T) {
+	var err error
+	var conn *Connection
+
+	conn, err = Connect(server, opts)
+	if err != nil {
+		t.Fatalf("Failed to connect: %s", err.Error())
+		return
+	}
+	if conn == nil {
+		t.Fatalf("conn is nil after Connect")
+		return
+	}
+	defer conn.Close()
+
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now())
+	defer cancel()
+	_, err = conn.CallAsync("simple_sleep", []interface{}{1}).WithCtx(ctx).Get()
+	if err != nil && err.Error() != "context is done" {
+		t.Error("context is not canceled", err.Error())
+	}
+
+	ctx, cancel = context.WithCancel(context.Background())
+	cancel()
+	_, err = conn.CallAsync("simple_sleep", []interface{}{1}).WithCtx(ctx).Get()
+	if err.Error() != "context is done" {
+		t.Error("context is not canceled", err.Error())
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), time.Millisecond)
```

Comment:
The timeout that expires earlier will delete the Future object from the queue. I have added some comments in tests, but I'm not sure that it is enough.

Reply:
It seems to me very strange to mix it. It seems to me that there should be one of two behaviors: timeout or context. Otherwise, I do not see much advantages of a context per a request.

Reply:
Using a context allows us to cancel the request not only with some timeout, but also by the deadline or with some cancel function that may be called conditionally.

Reply:
And so it will be very unexpectedly to cancel requests with context by a Connection timeout setting.

Reply:
Well, know I have splitted it into two requests. The methods DoWithContext, DoAsyncWithContext and DoTypedWith Context put the passed context to a future and if the timeouts loop sees that the future, that has an expired timeout, has a context not nil, it ignores it and doesn't remove from the queue. The alternative way is to give all futures with contexts a very big timeout, this way we don't need to patch the logic inside the timeouts() function.